### ISSUE TITLE
gprpc-swift tests stuck (docker)

### ISSUE THREAD

**Issue by weissi** [grpc/grpc-swift#430]
### New Issue Checklist

- [x] Updated grpc-swift to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/grpc/grpc-swift/blob/master/CONTRIBUTING.md)
- [x] I searched for [existing GitHub issues](https://github.com/grpc/grpc-swift/issues)(especially, there are many [resolved issues related to SSL](https://github.com/grpc/grpc-swift/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+ssl))

### Issue Description

as part of swift-nio-release tools, I run the grpc-swift tests in various configurations. The docker run has been stuck for a loooong time. grpc commit is 283754e80bb30a630e5e06508af654bc1d68fd7d . 

It seems like the promised allocated in  `file="/Users/johannes/.Trash/.nio-release-tools_qP5z8N/grpc-swift/Tests/SwiftGRPCNIOTests/ServerThrowingTests.swift", line=152` is not being fulfilled.

```
(lldb) bt
* thread #1, name = 'SwiftGRPCPackag', stop reason = signal SIGSTOP
  * frame #0: 0x00007f5a2ff41404 libpthread.so.0`__pthread_cond_wait + 196
    frame #1: 0x000055b2b10d32fa SwiftGRPCPackageTests.xctest`ConditionLock.lock(wantedValue=1, self=<unavailable>) at lock.swift:143:23
    frame #2: 0x000055b2b1013454 SwiftGRPCPackageTests.xctest`EventLoopFuture.wait(file="/Users/johannes/.Trash/.nio-release-tools_qP5z8N/grpc-swift/Tests/SwiftGRPCNIOTests/ServerThrowingTests.swift", line=152, self=<unavailable>) at EventLoopFuture.swift:873:14
    frame #3: 0x000055b2b13bba4c SwiftGRPCPackageTests.xctest`implicit closure #1 in ServerThrowingTests.testBidirectionalStreaming() at <compiler-generated>:0
    frame #4: 0x000055b2b13bbac6 SwiftGRPCPackageTests.xctest`partial apply for implicit closure #1 in ServerThrowingTests.testBidirectionalStreaming() at <compiler-generated>:0
    frame #5: 0x000055b2b0f66d56 SwiftGRPCPackageTests.xctest`thunk for @callee_guaranteed () -> (@error @owned Error) at <compiler-generated>:0
    frame #6: 0x000055b2b13bbb0b SwiftGRPCPackageTests.xctest`thunk for @callee_guaranteed () -> (@error @owned Error)partial apply at <compiler-generated>:0
    frame #7: 0x00007f5a30240d38 libXCTest.so`closure #1 () -> XCTest.(_XCTAssertionResult in _F1DE58039960C6E2C9E25CD38F016B36) in XCTest.XCTAssertNoThrow<A>(_: @autoclosure () throws -> A, _: @autoclosure () -> Swift.String, file: Swift.StaticString, line: Swift.UInt) -> () + 56
    frame #8: 0x00007f5a30240a58 libXCTest.so`XCTest.XCTAssertNoThrow<A>(_: @autoclosure () throws -> A, _: @autoclosure () -> Swift.String, file: Swift.StaticString, line: Swift.UInt) -> () + 40
    frame #9: 0x000055b2b13bb561 SwiftGRPCPackageTests.xctest`ServerThrowingTests.testBidirectionalStreaming(self=<unavailable>) at ServerThrowingTests.swift:152:5
    frame #10: 0x000055b2b13c1ec9 SwiftGRPCPackageTests.xctest`partial apply for ServerThrowingTests.testBidirectionalStreaming() at <compiler-generated>:0
    frame #11: 0x000055b2b0f66f86 SwiftGRPCPackageTests.xctest`thunk for @escaping @callee_guaranteed () -> (@error @owned Error) at <compiler-generated>:0
    frame #12: 0x000055b2b13c5dab SwiftGRPCPackageTests.xctest`thunk for @escaping @callee_guaranteed () -> (@error @owned Error)partial apply at <compiler-generated>:0
    frame #13: 0x00007f5a30231951 libXCTest.so`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed () -> (@out (), @error @owned Swift.Error) to @escaping @callee_guaranteed () -> (@error @owned Swift.Error) + 17
    frame #14: 0x00007f5a302316e8 libXCTest.so`partial apply forwarder for closure #1 (XCTest.XCTestCase) throws -> () in XCTest.(test in _3BE257A46ADB477C7BF2D39968B39F9D)<A where A: XCTest.XCTestCase>((A) -> () throws -> ()) -> (XCTest.XCTestCase) throws -> () + 88
    frame #15: 0x00007f5a30231654 libXCTest.so`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed XCTest.XCTestCase) -> (@error @owned Swift.Error) to @escaping @callee_guaranteed (@in_guaranteed XCTest.XCTestCase) -> (@out (), @error @owned Swift.Error) + 20
    frame #16: 0x00007f5a302319b9 libXCTest.so`reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed XCTest.XCTestCase) -> (@error @owned Swift.Error) to @escaping @callee_guaranteed (@in_guaranteed XCTest.XCTestCase) -> (@out (), @error @owned Swift.Error)partial apply forwarder with unmangled suffix ".37" + 9
    frame #17: 0x00007f5a3021e257 libXCTest.so`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@in_guaranteed XCTest.XCTestCase) -> (@out (), @error @owned Swift.Error) to @escaping @callee_guaranteed (@guaranteed XCTest.XCTestCase) -> (@error @owned Swift.Error) + 39
    frame #18: 0x00007f5a3022f76d libXCTest.so`XCTest.XCTestCase.invokeTest() -> () + 77
    frame #19: 0x00007f5a3022f34c libXCTest.so`XCTest.XCTestCase.perform(XCTest.XCTestRun) -> () + 172
    frame #20: 0x00007f5a302353f2 libXCTest.so`XCTest.XCTest.run() -> () + 194
    frame #21: 0x00007f5a30231dfe libXCTest.so`XCTest.XCTestSuite.perform(XCTest.XCTestRun) -> () + 254
    frame #22: 0x00007f5a302353f2 libXCTest.so`XCTest.XCTest.run() -> () + 194
    frame #23: 0x00007f5a30231dfe libXCTest.so`XCTest.XCTestSuite.perform(XCTest.XCTestRun) -> () + 254
    frame #24: 0x00007f5a302353f2 libXCTest.so`XCTest.XCTest.run() -> () + 194
    frame #25: 0x00007f5a30231e20 libXCTest.so`XCTest.XCTestSuite.perform(XCTest.XCTestRun) -> () + 288
    frame #26: 0x00007f5a302353f2 libXCTest.so`XCTest.XCTest.run() -> () + 194
    frame #27: 0x00007f5a3022d9d4 libXCTest.so`XCTest.XCTMain(Swift.Array<(testCaseClass: XCTest.XCTestCase.Type, allTests: Swift.Array<(Swift.String, (XCTest.XCTestCase) throws -> ())>)>) -> Swift.Never + 1652
    frame #28: 0x000055b2b13cc21d SwiftGRPCPackageTests.xctest`main at LinuxMain.swift:10:1
    frame #29: 0x00007f5a2d4f9f45 libc.so.6`__libc_start_main + 245
    frame #30: 0x000055b2b0c95c04 SwiftGRPCPackageTests.xctest`_start + 41
```

package.resolved:

```
{
  "object": {
    "pins": [
      {
        "package": "Commander",
        "repositoryURL": "https://github.com/kylef/Commander.git",
        "state": {
          "branch": null,
          "revision": "e5b50ad7b2e91eeb828393e89b03577b16be7db9",
          "version": "0.8.0"
        }
      },
      {
        "package": "Spectre",
        "repositoryURL": "https://github.com/kylef/Spectre.git",
        "state": {
          "branch": null,
          "revision": "f14ff47f45642aa5703900980b014c2e9394b6e5",
          "version": "0.9.0"
        }
      },
      {
        "package": "swift-nio-ssl-support",
        "repositoryURL": "https://github.com/apple/swift-nio-ssl-support.git",
        "state": {
          "branch": null,
          "revision": "c02eec4e0e6d351cd092938cf44195a8e669f555",
          "version": "1.0.0"
        }
      },
      {
        "package": "swift-nio-zlib-support",
        "repositoryURL": "https://github.com/apple/swift-nio-zlib-support.git",
        "state": {
          "branch": null,
          "revision": "37760e9a52030bb9011972c5213c3350fa9d41fd",
          "version": "1.0.0"
        }
      },
      {
        "package": "SwiftProtobuf",
        "repositoryURL": "https://github.com/apple/swift-protobuf.git",
        "state": {
          "branch": null,
          "revision": "6520fb185db88c0774a929acea1f7d5981a30d3a",
          "version": "1.4.0"
        }
      }
    ]
  },
  "version": 1
}
```

NIO repositories manually pinned to

```
swift-nio: e32a436466b01d4d45555356eb5bbdcdf14b4526
swift-nio-http2: 08db9d8cd4ec5dcb36787f1f7341418cd8c4fa97
swift-nio-ssl: 4f9d64ddd04cd1fe97d74dad7b5a77b882d8bf5d
```

this is Swift 5.0, on docker (ubuntu 14.04)

**Comment by weissi** [comment]
CC @glbrntt / @MrMage

**Comment by weissi** [comment]
all threads backtrace attached. Btw, there seems to be a thread leak, we have 118 threads running. Are you always calling `syncShutdownGracefully()` on the NIO `EventLoopGroup`s?

[bt-all.txt](https://github.com/grpc/grpc-swift/files/3081508/bt-all.txt)

**Comment by MrMage** [comment]
Thanks for the report! Would you mind providing the exact command you use to run the tests inside Docker, and maybe the corresponding `lldb` invocation as well (I am a noob with regards to debugging outside Xcode)?

**Comment by MrMage** [comment]
Update: the event loop were actually closed correctly, but spun up too _early_. This will be fixed in https://github.com/grpc/grpc-swift/pull/431 with https://github.com/grpc/grpc-swift/pull/431/commits/ca691e1d69a076a31392c6b01f47176c91b86a96.

**Comment by weissi** [comment]
> Thanks for the report! Would you mind providing the exact command you use to run the tests inside Docker

just `swift test`

>  and maybe the corresponding lldb invocation as well (I am a noob with regards to debugging outside Xcode)?

Sure, when I noticed that it was stuck, I just `docker exec -it my_running_container_id bash` and in there I took the pid of the running `*.xctest` binary and then did `lldb -p the_xctest_binary_pid` and then I did `bt` and `bt all` for the backtrace of the current thread and all the threads.

**Comment by MrMage** [comment]
> just `swift test`

Sure, but which Docker invocation do you use, in particular what is the base image?

**Comment by weissi** [comment]
@MrMage sorry, it was an internal Swift 5.0 one. I use

```
    docker run -it --rm --privileged -w "$PWD" -v "$working_dir:$working_dir" \
        "$image" "$@"
``` 

and image I build off something relatively close to the official dockers but with

```
cat >> "$dockerfile" <<EOF
FROM swift:5.0
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update; apt-get -y install lsof libssl-dev
EOF
```

on top. So `docker build -t "$nio_release_tools_docker_image" - < "$dockerfile" > /dev/null` where `$dockerfile` is the above snipped (modified to be `FROM swift:5.0` rather than `FROM internal_thing`)

**Comment by weissi** [comment]
ok, I can repro this 100% with the regular `swift:5.0` docker image. Here's the full trace:

```
johannes:~/.Trash/.nio-release-tools_WPakBo
$ docker run -it --rm  -v "$PWD:$PWD" -w "$PWD" swift:5.0 bash
root@71109ea3bb0c:/Users/johannes/.Trash/.nio-release-tools_WPakBo# cd grpc-swift/
root@71109ea3bb0c:/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift# swift test
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/JSONScanner.swift:192:15: warning: 'withUnsafeMutableBytes' is deprecated: use `withUnsafeMutableBytes<R>(_: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R` instead
    try value.withUnsafeMutableBytes {
              ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/JSONScanner.swift:850:25: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
        let n = try raw.withUnsafeBytes {
                        ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/JSONScanner.swift:909:25: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
        let n = try raw.withUnsafeBytes {
                        ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/JSONScanner.swift:973:27: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
          let n = try raw.withUnsafeBytes {
                          ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/JSONScanner.swift:1044:27: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
          let n = try raw.withUnsafeBytes {
                          ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/Message+BinaryAdditions.swift:36:14: warning: 'withUnsafeMutableBytes' is deprecated: use `withUnsafeMutableBytes<R>(_: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R` instead
    try data.withUnsafeMutableBytes { (pointer: UnsafeMutablePointer<UInt8>) in
             ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/Message+BinaryAdditions.swift:107:16: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
      try data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
               ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/Message+JSONAdditions.swift:92:22: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
    try jsonUTF8Data.withUnsafeBytes { (bytes:UnsafePointer<UInt8>) in
                     ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/Message+JSONArrayAdditions.swift:94:29: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
    return try jsonUTF8Data.withUnsafeBytes { (bytes:UnsafePointer<UInt8>) in
                            ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/Message+TextFormatAdditions.swift:56:18: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
        try data.withUnsafeBytes { (bytes: UnsafePointer<UInt8>) in
                 ^
tProtobuf/TextFormatEncoder.swift:266:15: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
        value.withUnsafeBytes { (p: UnsafePointer<UInt8>) in
              ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/AnyMessageStorage.swift:54:16: warning: 'init(bytes:)' is deprecated: use `init(_:)` instead
  var result = Data(bytes: [asciiOpenCurlyBracket])
               ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/AnyMessageStorage.swift:69:19: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
  try contentJSON.withUnsafeBytes { (bytes:UnsafePointer<UInt8>) in
                  ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/BinaryDecoder.swift:879:23: warning: 'withUnsafeMutableBytes' is deprecated: use `withUnsafeMutableBytes<R>(_: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R` instead
                field.withUnsafeMutableBytes { (pointer: UnsafeMutablePointer<UInt8>) in
                      ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/BinaryDecoder.swift:1207:30: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
                    try data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
                             ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/BinaryDecoder.swift:1248:33: warning: 'withUnsafeMutableBytes' is deprecated: use `withUnsafeMutableBytes<R>(_: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R` instead
                        payload.withUnsafeMutableBytes { (pointer: UnsafeMutablePointer<UInt8>) in
                                ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/BinaryDelimited.swift:58:10: warning: 'withUnsafeMutableBytes' is deprecated: use `withUnsafeMutableBytes<R>(_: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R` instead
    data.withUnsafeMutableBytes { (pointer: UnsafeMutablePointer<UInt8>) in
         ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/BinaryDelimited.swift:64:10: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
    data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
         ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/BinaryDelimited.swift:157:10: warning: 'withUnsafeMutableBytes' is deprecated: use `withUnsafeMutableBytes<R>(_: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R` instead
    data.withUnsafeMutableBytes { (pointer: UnsafeMutablePointer<UInt8>) in
         ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/Google_Protobuf_Any+Extensions.swift:69:18: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
        try data.withUnsafeBytes { (bytes: UnsafePointer<UInt8>) in
                 ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/JSONEncoder.swift:73:44: warning: 'init(bytes:)' is deprecated: use `init(_:)` instead
    internal var dataResult: Data { return Data(bytes: data) }
                                           ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/JSONEncoder.swift:336:19: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
            value.withUnsafeBytes { (p: UnsafePointer<UInt8>) in
                  ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift:71:17: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
      try bytes.withUnsafeBytes { (p: UnsafePointer<UInt8>) -> () in
                ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift:100:21: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
              bytes.withUnsafeBytes { (p: UnsafePointer<UInt8>) -> () in
                    ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/SwiftProtobuf/TextFormatScanner.swift:358:12: warning: 'withUnsafeMutableBytes' is deprecated: use `withUnsafeMutableBytes<R>(_: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R` instead
      data.withUnsafeMutableBytes {
           ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/protoc-gen-swift/FileIo.swift:50:11: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
    bytes.withUnsafeBytes { (p: UnsafePointer<UInt8>) -> () in
          ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/.build/checkouts/swift-protobuf/Sources/protoc-gen-swift/FileIo.swift:71:16: warning: 'init(bytes:)' is deprecated: use `init(_:)` instead
        return Data(bytes: buff)
               ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/Sources/protoc-gen-swiftgrpc/io.swift:49:16: warning: 'init(bytes:)' is deprecated: use `init(_:)` instead
        return Data(bytes: buff)
               ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/Sources/protoc-gen-swiftgrpc/io.swift:58:11: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
    bytes.withUnsafeBytes { (p: UnsafePointer<UInt8>) -> Void in
          ^
/Users/johannes/.Trash/.nio-release-tools_WPakBo/grpc-swift/Sources/SwiftGRPC/Core/ByteBuffer.swift:39:10: warning: 'withUnsafeBytes' is deprecated: use `withUnsafeBytes<R>(_: (UnsafeRawBufferPointer) throws -> R) rethrows -> R` instead
    data.withUnsafeBytes { bytes in
         ^
[1138/1138] Compiling BoringSSL crypto/asn1/a_bitstr.c
Test Suite 'All tests' started at 2019-04-17 09:57:41.800
Test Suite 'debug.xctest' started at 2019-04-17 09:57:41.825
Test Suite 'ClientThrowingWhenServerReturningErrorTests' started at 2019-04-17 09:57:41.825
Test Case 'ClientThrowingWhenServerReturningErrorTests.testBidirectionalStreaming' started at 2019-04-17 09:57:41.825

[stuck]
```

**Comment by MrMage** [comment]
I was able to reproduce the issue. It seems to be related with Docker's networking, as changing `host/hostname` for both client and server in https://github.com/grpc/grpc-swift/blob/bf4bb07d7d7e851136582e7cf697a3bd0ee2df11/Tests/SwiftGRPCNIOTests/NIOBasicEchoTestCase.swift#L112 to `127.0.0.1` makes the first few tests pass. At that point, however, our SSL tests fail:

```
Test Case 'NIOFunctionalTestsAnonymousClient.testBidirectionalStreamingBatched' started at 2019-04-17 10:47:55.100
Fatal error: 'try!' expression unexpectedly raised an error: NIO.NIOConnectionError(host: "127.0.0.1", port: 5050, dnsAError: nil, dnsAAAAError: nil, connectionErrors: [NIO.SingleConnectionFailure(target: [IPv4]127.0.0.1/127.0.0.1:5050, error: NIOSSL.BoringSSLError.invalidSNIName([]))]): file /Users/daniel/Documents/Xcode/Mac/grpc-swift/Tests/SwiftGRPCNIOTests/NIOBasicEchoTestCase.swift, line 146
```

Appending `hostOverride: "localhost"` yields a different error:

```
Test Case 'NIOFunctionalTestsAnonymousClient.testBidirectionalStreamingBatched' started at 2019-04-17 10:50:54.547
starting server 1
starting server 5
starting server 2
starting server 4
Fatal error: Unexpectedly found nil while unwrapping an Optional value
Current stack trace:
0    libswiftCore.so                    0x00007f888e6fe920 _swift_stdlib_reportFatalError + 69
1    libswiftCore.so                    0x00007f888e60ca06 <unavailable> + 3279366
2    libswiftCore.so                    0x00007f888e60cd85 <unavailable> + 3280261
3    libswiftCore.so                    0x00007f888e434810 _fatalErrorMessage(_:_:file:line:flags:) + 19
```

**Comment by weissi** [comment]
I can also confirm that removing `::1    localhost` from `/etc/hosts` does fix the issue. So it seems like it's not working on IPv6.

**Comment by MrMage** [comment]
Weirdly enough, establishing the connection in https://github.com/grpc/grpc-swift/blob/bf4bb07d7d7e851136582e7cf697a3bd0ee2df11/Sources/SwiftGRPCNIO/GRPCClientConnection.swift#L35 and `wait`ing on that succeeds, even though no connection is being established. (Or is establishing the connection postponed until it is actually used when no SSL handler is being present?)

So we have two issues here:
1. `GRPCClientConnection.start` should fail when it can't establish a connection.
2. We should figure out why the connection fails when IPv6 is involved.
3. We should find an easy way of forcing use of IPv4 (provided the issue is simply Docker's networking) while avoiding the issues outlined in https://github.com/grpc/grpc-swift/issues/430#issuecomment-484036544.

**Comment by glbrntt** [comment]
> 1. GRPCClientConnection.start should fail when it can't establish a connection.

I actually have a fix for this locally. Working on some tests around TLS handshake failures.

**Comment by glbrntt** [comment]
Ah actually I'm pretty sure I know what is causing the hang and think I have already resolved it locally (I came across this during interop testing). 

When we establish the client connection we're fulfilling a promise for the multiplexer in the channel initialiser. We can only fulfil this promise once but Happy Eyeballs will run it for IPv4 and IPv6. I.e. we can end up with a multiplexer for the wrong channel.

I'll try the local patch and report back...
